lib/src/main/scala/spinal/lib/com/jtag/sim/JtagDriver.scala

```diff
@@ -0,0 +1,59 @@
+package spinal.lib.com.jtag.sim
+
+import scala.collection.mutable
+import spinal.core.TimeNumber
+import spinal.core.sim._
+import spinal.lib.com.jtag.Jtag
+
+/*
+ * Jtag Driver tool
+ * Author: Alexis Marquet
+ */
+case class JtagDriver(jtag: Jtag, clockPeriod: TimeNumber) {
+
+  /*
+   * Reset the JTAG tap into Test-Logic-Reset state
+   */
+  def doResetTap(): Unit = {
+    doTmsSeq(Seq(true, true, true, true, true))
+  }
+
+  /*
+   * Send a sequence of TMS bits to move the JTAG tap into a specific state
+   */
+  def doTmsSeq(tmsSeq: Seq[Boolean]): Unit = {
+    for (tms <- tmsSeq) {
+      jtag.tms #= tms
+      doClockCycles(1)
+    }
+  }
+
+  /*
+   * Send n clock cycle to the JTAG tap
+   */
+  def doClockCycles(n: Long): Unit = {
+    for (_ <- 0 until n.toInt) {
+      jtag.tck #= true
+      sleep(timeToLong(clockPeriod / 2))
+      jtag.tck #= false
+      sleep(timeToLong(clockPeriod / 2))
+    }
+  }
+
+  /*
+   * Send a sequence of TDI bits to the JTAG tap and return the TDO sequence
+   * param: flipTms is used to flip the last TMS bit to 1 when the scan chain is finished
+   * This allows to write longer sequences of bits from multiple calls to this function
+   */
+  def doScanChain(tdiSeq: Seq[Boolean], flipTms: Boolean): Seq[Boolean] = {
+    // use mutable seq
+    val tdoSeq = mutable.Seq.fill(tdiSeq.length)(false)
+    for (i <- tdiSeq.indices) {
+      jtag.tdi #= tdiSeq(i)
+      jtag.tms #= (i == tdiSeq.length - 1 && flipTms)
+      tdoSeq(i) = jtag.tdo.toBoolean
+      doClockCycles(1)
+    }
+    tdoSeq.to[collection.immutable.Seq]
```

Comment:
This can't work since the type parameters for Seq are missing
```suggestion
    tdoSeq.toSeq
```
See also https://scastie.scala-lang.org/t1vaui2CR1qdlCtpZDfAFA

Reply:
that's weird, I do not get a warning on my setup, but I do get one with `toSeq` as it gives me a "remove redundant collection conversion"... would it make sense then to juste return tdoSeq?